### PR TITLE
docstring updates, attempt to make `shape`/`len` more precise

### PR DESCRIPTION
**Pull request by ryan-williams**

### CODE REVIEW

File: src/tiledbsoma_ml/pytorch.py [hunk 1]
```diff
@@ -318,25 +314,31 @@ def __len__(self) -> int:
...
-        world_size, _ = _get_distributed_world_rank()
-        n_workers, _ = _get_worker_world_rank()
-        partition_len = len(self._obs_joinids) // world_size // n_workers
-        div, rem = divmod(partition_len, self.batch_size)
-        return div + bool(rem), len(self._var_joinids)
+        world_size, rank = _get_distributed_world_rank()
+        n_workers, worker_id = _get_worker_world_rank()
+        obs_per_proc, obs_rem = divmod(len(self._obs_joinids), world_size)
+        # obs rows assigned to this "distributed" process
+        n_proc_obs = obs_per_proc + bool(rank < obs_rem)
```
Comment (bkmartinjr) [inline src/tiledbsoma_ml/pytorch.py:335]: I believe this is incorrect.  Every GPU gets the same number of samples (this is a hard requirement).  Counts can vary across multiple data loader workers, but each GPU worker must have exactly the same sample count.  See notes in `_create_obs_joinids_partition`.

If the partitioning across GPUs does not have the same carddinatlity, you get a crash or stall when using DDP.

Comment (ryan-williams) [reply src/tiledbsoma_ml/pytorch.py]: Thanks for explaining, I've updated it to reflect that "distributed" processes get rounded-down splits, but their child "data-loader" processes can have ±0.5.

I'll add some simple tests of this math as well, later. I think it's worth codifying our assumptions/intentions here.

**Review by bkmartinjr — APPROVED** [review]

File: src/tiledbsoma_ml/pytorch.py [hunk 2]
```diff
@@ -438,13 +438,16 @@ def _mini_batch_iter(
                         if self.return_sparse_X
                         else X_io_batch[iob_idx : iob_idx + mini_batch_size].toarray()
                     )
+                    obs_batch = obs_io_batch.iloc[iob_idx : iob_idx + mini_batch_size]
+                    # Give `obs_batch` a fresh [0, N) index (instead of an out-of-context slice of the IO batch's index)
+                    obs_batch.index = pd.RangeIndex(len(obs_batch))
```
Comment (bkmartinjr) [inline src/tiledbsoma_ml/pytorch.py:443]: why not the (slightly) more readable:
> obs_batch = obs_io_batch.iloc[iob_idx : iob_idx + mini_batch_size].reset_index(drop=True)

I'm fine w/ either, but this seems much more explicit (to the point that the comment isn't really necessary)

Comment (ryan-williams) [reply src/tiledbsoma_ml/pytorch.py]: Agreed, done.

**Review by bkmartinjr — APPROVED** [review]
Approving, but PTAL at the style/clarity suggestion in pytorch.py

**Pull request merged**